README.md
# memoredis

Redis memoization library with good Typescript generics, locking, and argument-wide bulk invalidation

# Usage

```
yarn add memoredis
```

## Initialization

```typescript
import memoredis from 'memoredis';

// pass options as they would directly to redis.createClient
const memoizer = memoredis({ clientOpts: 'redis://' });

// or pass in an existing client
const client = redis.createClient();
const memoizer = memoredis({ client });
```

## Basic Usage

```typescript
const expensiveDatabaseLookup = async ({ authorId, genre }) => {
  // some expensive database lookup
  return { countOfBooks: n };
};

const countOfBooksPublishedByAuthorInGenre = memoizer.memoize(
  expensiveDatabaseLookup,
  { key: 'genre-author-count', ttl: 60 * 1000 }
);

await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // goes to the database
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // served from redis!
```

## Basic Invalidation

```typescript
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // goes to the database
await memoizer.invalidate('genre-author-count', {
  authorId: 4,
  genre: 'fiction'
});
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // goes to the database
```

## Bulk Partial Invalidation

```typescript
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // goes to the database
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'biology' }); // goes to the database
await memoizer.invalidate('genre-author-count', {
  authorId: 4
});
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'fiction' }); // goes to the database
await countOfBooksPublishedByAuthorInGenre({ authorId: 4, genre: 'biology' }); // goes to the database
```

## Prefixes

```typescript
// optionally include a prefix for namespacing caches
const libraryA = memoredis({ prefix: 'a' });
const libraryB = memoredis({ prefix: 'b' });

const expensiveDatabaseLookup = async ({ authorId, genre }) => {
  // some expensive database lookup
  console.log('Finished looking up');
  return { countOfBooks: n };
};

const countOfBooksPublishedByAuthorInGenreInLibraryA = libraryA.memoize(
  expensiveDatabaseLookup,
  { key: 'genre-author-count', ttl: 60 * 1000 }
);

const countOfBooksPublishedByAuthorInGenreInLibraryB = libraryA.memoize(
  expensiveDatabaseLookup,
  { key: 'genre-author-count', ttl: 60 * 1000 }
);

await countOfBooksPublishedByAuthorInGenreInLibraryA({
  authorId: 4,
  genre: 'fiction'
}); // goes to the database

await countOfBooksPublishedByAuthorInGenreInLibraryB({
  authorId: 4,
  genre: 'fiction'
}); // goes to the database
```

## Locking

This library uses `redis-lock` under the hood to prevent two calls to the same function
at the same time, even when the cache is cold.

```typescript
const expensiveDatabaseLookup = async ({ authorId, genre }) => {
  // some expensive database lookup
  console.log('Finished looking up');
  return { countOfBooks: n };
};

const countOfBooksPublishedByAuthorInGenre = memoizer.memoize(
  expensiveDatabaseLookup,
  { key: 'genre-author-count', ttl: 60 * 1000 }
);

const args = { authorId: 4, genre: 'fiction' };
await Promise.all([
  countOfBooksPublishedByAuthorInGenre(args),
  countOfBooksPublishedByAuthorInGenre(args)
]);
// only logs 'Finished looking up' once
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/politics-rewired/memoredis/compare/v1.1.1...v2.0.0) (2022-10-21)


### Features

* update to redis v4 cli ([#33](https://github.com/politics-rewired/memoredis/issues/33)) ([8172642](https://github.com/politics-rewired/memoredis/commit/8172642b85bf231b75346821c65eb4f1377f60c7))

### [1.1.1](https://github.com/politics-rewired/memoredis/compare/v1.1.0...v1.1.1) (2021-01-26)

## [1.1.0](https://github.com/politics-rewired/memoredis/compare/v1.0.8...v1.1.0) (2020-09-07)


### Features

* expose redis client quit and end methods ([bc23715](https://github.com/politics-rewired/memoredis/commit/bc237156ae76661ba9dbc55a148b32a0dba7d845))


### Bug Fixes

* allow non-yarn package manager ([219939e](https://github.com/politics-rewired/memoredis/commit/219939e5deee7469206a418932065ccabbac50e0))

## [1.0.8](https://github.com/politics-rewired/memoredis/compare/v1.0.7...v1.0.8) (2020-02-14)


### Bug Fixes

* add key to invalidation list on fetch in case cache has been flushed ([2084734](https://github.com/politics-rewired/memoredis/commit/2084734b15befddc1cf45a1d11bef40b4af80ab8))
* move benchmark out of source to note compile ([b8171bd](https://github.com/politics-rewired/memoredis/commit/b8171bd87a78fd85cdc2fe550768d031e8e9f9eb))



## [1.0.7](https://github.com/politics-rewired/memoredis/compare/v1.0.6...v1.0.7) (2020-02-06)


### Features

* speed up cache invalidation using key sets ([56bd24a](https://github.com/politics-rewired/memoredis/commit/56bd24a9f516c5512243b5b314e7b88e872d3c23))



## [1.0.6](https://github.com/politics-rewired/memoredis/compare/v1.0.5...v1.0.6) (2020-01-30)


### Bug Fixes

* return null for errors in get ([75905b9](https://github.com/politics-rewired/memoredis/commit/75905b9f584d416bb9764ad2ed6b4247898ba139))



## [1.0.5](https://github.com/politics-rewired/memoredis/compare/v1.0.4...v1.0.5) (2020-01-21)


### Bug Fixes

* scan all pagination ([bae23da](https://github.com/politics-rewired/memoredis/commit/bae23dafe37d3cd1ffde92076d68cc566c905193))



## [1.0.4](https://github.com/politics-rewired/memoredis/compare/v1.0.3...v1.0.4) (2020-01-20)


### Bug Fixes

* deserialize dates ([5ef962a](https://github.com/politics-rewired/memoredis/commit/5ef962a70f62e2351ad3ff7d3b37e20365ac2d22))



## [1.0.3](https://github.com/politics-rewired/memoredis/compare/v1.0.2...v1.0.3) (2020-01-20)


### Features

* empty mode ([a2a9eb7](https://github.com/politics-rewired/memoredis/commit/a2a9eb77433d469bfb630e8e748096f1bb4e241c))
* use stable sha1 for nested objects ([62aabcf](https://github.com/politics-rewired/memoredis/commit/62aabcf40b2b49c144105c9d2e69cb058d6ca4e5))



## [1.0.2](https://github.com/politics-rewired/memoredis/compare/v1.0.1...v1.0.2) (2020-01-20)


### Bug Fixes

* **memoizer:** json deserialization ([1981209](https://github.com/politics-rewired/memoredis/commit/1981209ee6fdfdd67b05d5e4e3b74c0708af98d0))



## [1.0.1](https://github.com/politics-rewired/memoredis/compare/v1.0.0...v1.0.1) (2020-01-17)


### Bug Fixes

* proper index exports ([24439f6](https://github.com/politics-rewired/memoredis/commit/24439f6940e35ce013c6822e9d2be73c8b4ce4fc))



# [1.0.0](https://github.com/politics-rewired/memoredis/compare/99fd88847d803960779eb2628467afa21b00d3d4...v1.0.0) (2020-01-17)


### Features

* done, all tests passing ([99fd888](https://github.com/politics-rewired/memoredis/commit/99fd88847d803960779eb2628467afa21b00d3d4))


src/lib/memoizer.spec.ts
import { createClient } from 'redis';
import { createMemoizer, produceKeyWithArgs } from './memoizer';

const flushRedis = async () => {
  const client = createClient();
  await client.connect();
  await client.flushDb();
  await client.disconnect();
};

const sleep = (n: number) =>
  new Promise((resolve) => setTimeout(resolve, n * 1000));

describe('produceKeyWithArgs', () => {
  test('produceKeyWithArgs should produce the same key for differently ordered objects', () => {
    expect(produceKeyWithArgs('prefix', 'key', { a: 1, b: 2 })).toEqual(
      produceKeyWithArgs('prefix', 'key', { b: 2, a: 1 })
    );
  });

  test('should use sha1 for nested object values', () => {
    const keyHash = produceKeyWithArgs('prefix', 'key', { a: { b: 1 } });
    expect(keyHash).toMatch(/|prefix|key|a:.*/);
  });
});

describe('basic functionality', () => {
  afterAll(async () => {
    await flushRedis();
  });

  test('should only call function once for same args', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'basic',
    });

    const args = { a: 4 };
    await memoizedFunction(args);
    await memoizedFunction(args);
    await memoizer.quit();

    expect(mock).toHaveBeenCalledTimes(1);
  });

  test('should distinguish different args', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'different',
    });

    await memoizedFunction({ a: 1 });
    await memoizedFunction({ a: 2 });
    await memoizer.quit();

    expect(mock).toHaveBeenCalledTimes(2);
  });

  test('should lock duplicate calls', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableSlowFunction = async () => {
      await sleep(1);
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableSlowFunction, {
      key: 'locking',
    });

    await Promise.all([memoizedFunction({ a: 1 }), memoizedFunction({ a: 1 })]);
    await memoizer.quit();

    expect(mock).toHaveBeenCalledTimes(1);
  });
});

describe('invalidation', () => {
  afterAll(async () => {
    await flushRedis();
  });

  test('should call twice if invalidated', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'basic-invalidate',
    });

    const args = { a: 4 };
    await memoizedFunction(args);
    await memoizer.invalidate('basic-invalidate', args);
    await memoizedFunction(args);
    expect(mock).toHaveBeenCalledTimes(2);

    await memoizer.quit();
  });

  test('should call once if different arg is invalidated', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'missed-invalidate',
    });

    await memoizedFunction({ a: 4 });
    await memoizer.invalidate('basic-invalidate', { a: 5 });
    await memoizedFunction({ a: 4 });
    expect(mock).toHaveBeenCalledTimes(1);

    await memoizer.quit();
  });

  test('should call twice if different compound covered invalidate', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'compound-covered-invalidate',
    });

    await memoizedFunction({ a: 1, b: 2 });
    await memoizer.invalidate('compound-covered-invalidate', { a: 1 });
    await memoizedFunction({ a: 1, b: 2 });
    expect(mock).toHaveBeenCalledTimes(2);

    await memoizer.quit();
  });

  test('should call once if invalidation is partially wrong', async () => {
    const memoizer = await createMemoizer({});

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'compound-missed-invalidate',
    });

    await memoizedFunction({ a: 1, b: 2 });
    await memoizer.invalidate('compound-missed-invalidate', { a: 1, b: 3 });
    await memoizedFunction({ a: 1, b: 2 });
    expect(mock).toHaveBeenCalledTimes(1);

    await memoizer.quit();
  });
});

describe('prefixes', () => {
  afterAll(async () => {
    await flushRedis();
  });

  test('should keep keyspaces separate', async () => {
    const a = await createMemoizer({ prefix: 'a' });
    const b = await createMemoizer({ prefix: 'b' });

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const aF = a.memoize(memoizableFunction, {
      key: 'prefix-test',
    });

    const bF = b.memoize(memoizableFunction, {
      key: 'prefix-test',
    });

    await aF({ a: 1 });
    await bF({ a: 1 });

    expect(mock).toHaveBeenCalledTimes(2);

    await a.quit();
    await b.quit();
  });
});

describe('empty mode', () => {
  afterAll(async () => {
    await flushRedis();
  });

  test('if empty mode should not cache', async () => {
    const memoizer = await createMemoizer({ emptyMode: true });

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = memoizer.memoize(memoizableFunction, {
      key: 'empty-mode-test',
    });

    await memoizedFunction({ a: 1 });
    await memoizedFunction({ a: 1 });
    expect(mock).toHaveBeenCalledTimes(2);

    await memoizer.quit();
  });
});

describe('scan all cursor', () => {
  afterAll(async () => {
    await flushRedis();
  });

  test('if we have 1000 keys we should be able to invalidate them all quickly', async () => {
    const m = await createMemoizer({ prefix: 'a' });

    const mock = jest.fn();

    const memoizableFunction = async () => {
      mock();
      return 4;
    };

    const memoizedFunction = m.memoize(memoizableFunction, {
      key: 'scan-all-delete',
    });

    // Call it 1000 times with different args and one same arg
    await Promise.all(
      new Array(1000)
        .fill(null)
        .map((_, n) => memoizedFunction({ n, same: true }))
    );

    await m.invalidate('scan-all-delete', { same: true });
    await memoizedFunction({ n: 1, same: true });
    expect(mock).toHaveBeenCalledTimes(1001);

    await m.quit();
  });
});


src/lib/redis.ts
import { RedisClientType } from 'redis';
import { Lock } from 'redis-promise-lock';

export const makeEasyRedis = (client: RedisClientType) => {
  const lockerRoom = new Lock(client);

  const withLock = async (key: string, timeout: number, fn: () => void) => {
    try {
      await lockerRoom.acquireLock(`lock-${key}`, { ttl: timeout / 1000 });
      const result = await fn();
      return result;
    } finally {
      await lockerRoom.releaseLock(`lock-${key}`);
    }
  };

  const pSetEx = async (key: string, milliseconds: number, value: any) => {
    const jsonValue = JSON.stringify(value);

    return client.pSetEx(key, milliseconds, jsonValue);
  };

  const pSetExAndSAdd = async (
    setKey: string,
    key: string,
    milliseconds: number,
    value: any
  ) => {
    const jsonValue = JSON.stringify(value);

    return client
      .multi()
      .sAdd(setKey, key)
      .pSetEx(key, milliseconds, jsonValue)
      .exec();
  };

  const get = async (key: string) => {
    const result = await client.get(key);

    return result ? JSON.parse(result) : null;
  };

  const scan = async (match: string, cursor?: number) =>
    client.scan(cursor ?? 0, {
      MATCH: match,
    });

  const scanAll = async (match: string) => {
    let keys: string[] = [];
    let cursor;

    while (cursor !== 0) {
      const result = await scan(match, cursor);

      cursor = result.cursor;
      keys = keys.concat(result.keys);
    }

    return keys;
  };

  const scanSet = async (setKey: string, match: string, cursor?: number) =>
    client.sScan(setKey, cursor ?? 0, { MATCH: match });

  const scanSetAll = async (setKey: string, match: string) => {
    let keys: string[] = [];
    let cursor;

    while (cursor !== 0) {
      const result = await scanSet(setKey, match, cursor);

      cursor = result.cursor;
      keys = keys.concat(result.members);
    }

    return keys;
  };

  const del = async (keys: string[]) => client.del(keys);

  const delAndRem = async (setKey: string, valueKeys: string[]) =>
    client.multi().del(valueKeys).sRem(setKey, valueKeys).exec();

  const sAdd = async (setKey: string, valueKey: string) =>
    client.sAdd(setKey, valueKey);

  const sRem = async (setKey: string, valueKeys: string[]) =>
    client.sRem(setKey, valueKeys);

  return {
    del,
    delAndRem,
    get,
    pSetEx,
    pSetExAndSAdd,
    sAdd,
    sRem,
    scan,
    scanAll,
    scanSet,
    scanSetAll,
    withLock,
  };
};


src/lib/memoizer.ts
import crypto from 'crypto';
import jsonStableStringify from 'json-stable-stringify';
import { createClient, RedisClientOptions, RedisClientType } from 'redis';
import { String } from 'runtypes';
import { makeEasyRedis } from './redis';

const SafeString = String.withConstraint(
  (s) => !s.includes(':') && !s.includes('|')
);

const DEFAULT_TTL = 60 * 1000;
const DEFAULT_LOCK_TIMEOUT = 5 * 1000;

export interface MemoizerOpts {
  prefix?: string;
  client?: RedisClientType;
  clientOpts?: RedisClientOptions;
  emptyMode?: boolean;
}

export interface MemoizeOpts {
  key: string;
  lockTimeout?: number;
  ttl?: number;
}

export interface MemoizedFunctionArgs {
  [key: string]: any;
}

export type MemoizableFunction<T, U> = (args: T) => Promise<U>;

export interface Memoizer {
  invalidate(key: string, args: MemoizedFunctionArgs): Promise<void>;
  memoize<T, U>(
    fn: MemoizableFunction<T, U>,
    opts: MemoizeOpts
  ): MemoizableFunction<T, U>;
  quit(): Promise<void>;
  end(): Promise<void>;
}

export const createMemoizer = async (
  instanceOpts: MemoizerOpts
): Promise<Memoizer> => {
  if (instanceOpts.emptyMode) {
    return {
      end: () =>
        new Promise((resolve) => {
          resolve();
        }),
      invalidate: async (_key: string, _forArgs: MemoizedFunctionArgs) => {
        // do nothing
      },
      memoize: <T, U>(fn: MemoizableFunction<T, U>, _opts: MemoizeOpts) => {
        return async (args: T): Promise<U> => {
          return fn(args);
        };
      },
      quit: async () =>
        new Promise((resolve) => {
          resolve();
        }),
    };
  }

  const client: RedisClientType = (
    instanceOpts.client
      ? instanceOpts.client
      : createClient(instanceOpts.clientOpts)
  ) as RedisClientType;

  await client.connect();

  const { withLock, pSetExAndSAdd, get, scanSetAll, delAndRem } =
    makeEasyRedis(client);

  if (instanceOpts.prefix) {
    SafeString.check(instanceOpts.prefix);
  }

  const invalidate = async (key: string, forArgs: MemoizedFunctionArgs) => {
    const setKey = produceSetKey(instanceOpts.prefix, key);
    SafeString.check(key);

    const glob =
      [produceKey(instanceOpts.prefix, key), ...stringifyArgs(forArgs)].join(
        '*'
      ) + '*';

    const cachedKeys = await scanSetAll(setKey, glob);

    if (cachedKeys.length > 0) {
      await delAndRem(setKey, cachedKeys);
    }
  };

  const memoize = <T, U>(fn: MemoizableFunction<T, U>, opts: MemoizeOpts) => {
    SafeString.check(opts.key);

    return async (args: T): Promise<U> => {
      const redisKey = produceKeyWithArgs(instanceOpts.prefix, opts.key, args);
      const setKey = produceSetKey(instanceOpts.prefix, opts.key);

      // attempt early memoized return
      const foundResult = await get(redisKey);
      if (foundResult) {
        return foundResult;
      }

      return withLock(
        redisKey,
        opts.lockTimeout || DEFAULT_LOCK_TIMEOUT,
        async () => {
          const foundResultAfterLock = await get(redisKey);
          if (foundResultAfterLock) {
            return foundResultAfterLock;
          }

          const result = await fn(args);
          await pSetExAndSAdd(
            setKey,
            redisKey,
            opts.ttl || DEFAULT_TTL,
            result
          );
          return result;
        }
      ) as U;
    };
  };

  const quit = async () => client.quit();

  const end = async () => client.disconnect();

  return { memoize, invalidate, quit, end };
};

export const produceKeyWithArgs = (
  prefix: string,
  key: string,
  args: MemoizedFunctionArgs
) => `${produceKey(prefix, key)}|${stringifyArgs(args).join('|')}`;

const produceKey = (prefix: string, key: string) =>
  prefix ? `${prefix}|${key}` : key;

const produceSetKey = (prefix: string, key: string) =>
  prefix ? `${prefix}|${key}-keyset` : `${key}-keyset`;

const stringifyArgs = (args: MemoizedFunctionArgs) =>
  Object.keys(args)
    .map((key) => stringifyArg(key, args[key]))
    .sort();

const stringifyArg = (key: string, value: any) =>
  `${key}:${typeof value === 'object' ? objectHash(value) : value}`;

const objectHash = (obj: Record<string, any>) => sha1(jsonStableStringify(obj));

const sha1 = (str: string) =>
  crypto.createHmac('sha1', 'memo').update(str).digest('hex');
